plugins/permission-backend/src/service/router.ts

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  errorHandler,
+  PluginEndpointDiscovery,
+} from '@backstage/backend-common';
+import express, { Request, Response } from 'express';
+import Router from 'express-promise-router';
+import { Logger } from 'winston';
+import {
+  BackstageIdentity,
+  IdentityClient,
+} from '@backstage/plugin-auth-backend';
+import { ConflictError } from '@backstage/errors';
+import {
+  AuthorizeResult,
+  AuthorizeResponse,
+  AuthorizeRequest,
+  Identified,
+} from '@backstage/plugin-permission-common';
+import { PermissionPolicy } from '@backstage/plugin-permission-node';
+import { PermissionIntegrationClient } from './PermissionIntegrationClient';
+
+/**
+ * Options required when constructing a new {@link express#Router} using
+ * {@link createRouter}.
+ *
+ * @public
+ */
+export interface RouterOptions {
+  logger: Logger;
+  discovery: PluginEndpointDiscovery;
+  policy: PermissionPolicy;
+}
+
+const handleRequest = async (
+  { id, resourceRef, ...request }: Identified<AuthorizeRequest>,
+  user: BackstageIdentity | undefined,
+  policy: PermissionPolicy,
+  permissionIntegrationClient: PermissionIntegrationClient,
+  authHeader?: string,
+): Promise<Identified<AuthorizeResponse>> => {
+  const response = await policy.handle(request, user);
+
+  if (response.result === AuthorizeResult.CONDITIONAL) {
+    // Sanity check that any resource provided matches the one expected by the permission
+    if (request.permission.resourceType !== response.conditions.resourceType) {
+      throw new ConflictError(
+        `Invalid resource conditions returned from permission policy for permission ${request.permission.name}`,
+      );
+    }
+
+    if (resourceRef) {
+      return {
+        id,
+        ...(await permissionIntegrationClient.applyConditions(
+          {
+            resourceRef,
+            ...response.conditions,
+          },
+          authHeader,
+        )),
+      };
+    }
+
+    return {
+      id,
+      result: AuthorizeResult.CONDITIONAL,
+      // TODO(mtlewis): this .conditions.conditions situation is a bit awkward. I think it's
+      // worth exploring a bit of reorganization of the ConditionalPolicyResult type so that
+      // the naming of property chains like this makes a bit more sense.
```

Comment:
Propose we leave this todo in place for now - since reorganizing this type will probably touch permission-common, permission-node and permission-backend, it probably makes sense to address it once all three are merged.